### PR TITLE
Security

### PR DESCRIPTION
**Pull request by YevheniiVlasenko**

### CODE REVIEW

File: src/main/java/project/bookstore/model/Role.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+@Data
+@Table(name = "roles")
+public class Role {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @Column(nullable = false, unique = true)
+    @JdbcTypeCode(SqlTypes.VARCHAR)
+    private String name;
```
Comment (ponomvrenko) [inline src/main/java/project/bookstore/model/Role.java]: We've already created enum with our names, so you should replace String type

File: src/main/java/project/bookstore/model/User.java [hunk 2]
```diff
@@ -35,4 +45,46 @@ public class User {
 
     private String shippingAddress;
 
+    @Column(nullable = false)
```
Comment (ponomvrenko) [inline src/main/java/project/bookstore/model/User.java:48]: Define here a name for column

**Review by ponomvrenko — CHANGES_REQUESTED** [review]

File: hs_err_pid11760.log
Comment (d1sam) [inline hs_err_pid11760.log:1]: what is a file?

File: src/main/java/project/bookstore/controller/BookController.java [hunk 3]
```diff
@@ -33,25 +34,29 @@ public class BookController {
     private final BookMapper bookMapper;
 
     @GetMapping
+    @PreAuthorize("hasRole('ROLE_USER')")
```
Comment (d1sam) [inline src/main/java/project/bookstore/controller/BookController.java]:
```suggestion
    @PreAuthorize("hasAnyAuthority('ROLE_USER', 'ROLE_ADMIN')")
```

File: src/main/java/project/bookstore/controller/BookController.java [hunk 4]
```diff
@@ -33,25 +34,29 @@ public class BookController {
...
 
     @GetMapping
+    @PreAuthorize("hasRole('ROLE_USER')")
     @Operation(summary = "Get all books", description = "Get list of all present books")
     public List<BookDto> getAll(Pageable pageable) {
         return bookService.findAll(pageable);
     }
 
     @GetMapping("/{id}")
+    @PreAuthorize("hasRole('ROLE_USER')")
```
Comment (d1sam) [inline src/main/java/project/bookstore/controller/BookController.java]:
```suggestion
    @PreAuthorize("hasAnyAuthority('ROLE_USER', 'ROLE_ADMIN')")
```

File: src/main/java/project/bookstore/model/Role.java (hunk 1 above)
Comment (d1sam) [inline src/main/java/project/bookstore/model/Role.java]:
```suggestion
    private RoleName name;
```

File: src/main/java/project/bookstore/model/User.java [hunk 5]
```diff
@@ -35,4 +45,46 @@ public class User {
...
+    }
+
+    @Override
+    public boolean isAccountNonLocked() {
+        return true;
+    }
+
+    @Override
+    public boolean isCredentialsNonExpired() {
+        return true;
```
Comment (d1sam) [inline src/main/java/project/bookstore/model/User.java]:
```suggestion
        return !isDeleted;
```

File: src/main/java/project/bookstore/model/User.java [hunk 6]
```diff
@@ -35,4 +45,46 @@ public class User {
...
+    }
+
+    @Override
+    public boolean isAccountNonExpired() {
+        return true;
+    }
+
+    @Override
+    public boolean isAccountNonLocked() {
+        return true;
```
Comment (d1sam) [inline src/main/java/project/bookstore/model/User.java]:
```suggestion
        return !isDeleted;
```

File: src/main/java/project/bookstore/model/User.java [hunk 7]
```diff
@@ -35,4 +45,46 @@ public class User {
...
+    }
+
+    @Override
+    public String getUsername() {
+        return email;
+    }
+
+    @Override
+    public boolean isAccountNonExpired() {
+        return true;
```
Comment (d1sam) [inline src/main/java/project/bookstore/model/User.java]:
```suggestion
        return !isDeleted;
```

File: src/main/java/project/bookstore/repository/RoleRepository.java [hunk 8]
```diff
@@ -0,0 +1,7 @@
+package project.bookstore.repository;
+
+import org.springframework.data.jpa.repository.JpaRepository;
+import project.bookstore.model.Role;
+
```
Comment (d1sam) [inline src/main/java/project/bookstore/repository/RoleRepository.java:6]:
```suggestion
@Repository
```

File: src/main/java/project/bookstore/security/SecurityConfig.java [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+    @Bean
+    public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Exception {
+        return http
+                .cors(AbstractHttpConfigurer::disable)
+                .csrf(AbstractHttpConfigurer::disable)
+                .authorizeHttpRequests(
+                        auth -> auth
+                                .requestMatchers("/auth/**", "/registration").permitAll()
+                                .requestMatchers("/swagger-ui/**", "/v3/api-docs/**",
+                                        "/swagger-resources/**").permitAll()
```
Comment (d1sam) [inline src/main/java/project/bookstore/security/SecurityConfig.java]:
```suggestion
                                .requestMatchers("/auth/**", "/swagger-ui/**", "/v3/api-docs/**").permitAll()
```

File: src/main/resources/db/changelog/changes/03-create-roles-table.yaml [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+                  autoIncrement: true
+                  constraints:
+                    primaryKey: true
+                    nullable: false
+              - column:
+                  name: name
+                  type: VARCHAR(255)
+                  constraints:
+                    unique: true
+                    nullable: false
```
Comment (d1sam) [inline src/main/resources/db/changelog/changes/03-create-roles-table.yaml]: Please add empty line to the end of file

File: src/main/resources/db/changelog/changes/04-create-users-roles-table.yaml [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+                    nullable: false
+                    foreignKeyName: fk_user_id
+                    references: users(id)
+              - column:
+                  name: roles_id
+                  type: BIGINT
+                  constraints:
+                    nullable: false
+                    foreignKeyName: fk_role_id
+                    references: roles(id)
```
Comment (d1sam) [inline src/main/resources/db/changelog/changes/04-create-users-roles-table.yaml]: Please add empty line to the end of file

File: src/main/resources/db/changelog/changes/05-insert-admin-user-to-db.yaml [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+            columns:
+              - column:
+                  name: id
+                  valueNumeric: 1
+              - column:
+                  name: email
+                  value: admin@example.com
+              - column:
+                  name: password
+                  value: qwerty
```
Comment (d1sam) [inline src/main/resources/db/changelog/changes/05-insert-admin-user-to-db.yaml:17]: password not hashed

File: src/main/resources/db/changelog/changes/06-add-customer-user-to-db.yaml [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+            columns:
+              - column:
+                  name: id
+                  valueNumeric: 2
+              - column:
+                  name: email
+                  value: customer@example.com
+              - column:
+                  name: password
+                  value: qwerty
```
Comment (d1sam) [inline src/main/resources/db/changelog/changes/06-add-customer-user-to-db.yaml:17]: password not hashed

File: src/main/resources/db/changelog/changes/08-add-roles-to-roles-table.yaml [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+        - insert:
+            tableName: roles
+            columns:
+              - column:
+                  name: id
+                  valueNumeric: 2
+              - column:
+                  name: name
+                  value: USER
+
```
Comment (d1sam) [inline src/main/resources/db/changelog/changes/08-add-roles-to-roles-table.yaml]: only 1 empty line should be at the end of file

File: src/main/resources/db/changelog/changes/06-add-customer-user-to-db.yaml (hunk 13 above)
Comment (d1sam) [reply src/main/resources/db/changelog/changes/06-add-customer-user-to-db.yaml:17]: I see you hash password in the next changesets, but please be careful with passwords inside your changesets

File: src/main/resources/db/changelog/changes/05-insert-admin-user-to-db.yaml (hunk 12 above)
Comment (d1sam) [reply src/main/resources/db/changelog/changes/05-insert-admin-user-to-db.yaml:17]: I see you hash password in the next changesets, but please be careful with passwords inside your changesets

File: src/main/resources/db/changelog/changes/09-add-role-column-to-user-table.yaml [hunk 15]
```diff
@@ -0,0 +1,11 @@
...
+  - changeSet:
+      id: 9
+      author: vlasenko
+      changes:
+        - addColumn:
+            tableName: users
+            columns:
+              - column:
+                  name: role
+                  type: VARCHAR(255)
```
Comment (d1sam) [inline src/main/resources/db/changelog/changes/09-add-role-column-to-user-table.yaml]: please add empty line to the end of file

Comment (d1sam) [reply src/main/resources/db/changelog/changes/09-add-role-column-to-user-table.yaml]: why do you need this column? you already have `users_roles` table

**Review by d1sam — CHANGES_REQUESTED** [review]
Good job, but some changes should be done according to my comment

File: hs_err_pid11760.log [hunk 16]
```diff
@@ -0,0 +1,1368 @@
+#
```
Comment (yevheniisynytsia) [inline hs_err_pid11760.log]: can you remove this file?

**Review by yevheniisynytsia — CHANGES_REQUESTED** [review]

File: src/main/resources/db/changelog/changes/04-create-users-roles-table.yaml
Comment (d1sam) [inline src/main/resources/db/changelog/changes/04-create-users-roles-table.yaml:1]: You should add **composite** [key](https://www.simplilearn.com/tutorials/sql-tutorial/composite-key-in-sql#:~:text=View%20More-,What%20Is%20a%20Composite%20Key%20in%20SQL%3F,can%20uniquely%20identify%20any%20record.)

File: src/main/resources/db/changelog/changes/10-assign-roles-to-existing-users.yaml
Comment (d1sam) [inline src/main/resources/db/changelog/changes/10-assign-roles-to-existing-users.yaml:1]: try to improve your code and use `valueComputed` with `preConditions` for your insert.
Example:
```
      preConditions:
        - sqlCheck:
            expectedResult: 1
            sql: SELECT COUNT(*) FROM users WHERE email='admin@example.com'
        - sqlCheck:
            expectedResult: 1
            sql: SELECT COUNT(*) FROM roles WHERE name='ROLE_ADMIN'
      changes:
        - insert:
            tableName: user_role
            columns:
              - column:
                  name: user_id
                  valueComputed: (SELECT id FROM users WHERE email='admin@example.com')
              - column:
                  name: role_id
                  valueComputed: (SELECT id FROM roles WHERE name='ROLE_ADMIN')
```

**Review by d1sam — CHANGES_REQUESTED** [review]
good job, but some changes in your migration files should be done.

File: src/main/resources/db/changelog/changes/10-assign-roles-to-existing-users.yaml
Comment (YevheniiVlasenko) [reply src/main/resources/db/changelog/changes/10-assign-roles-to-existing-users.yaml:1]: Isn't it just the same action with a lot of complications in this particular case?

File: src/main/resources/db/changelog/changes/10-add-primary-key-for-users-roles.yaml [hunk 17]
```diff
@@ -0,0 +1,9 @@
+databaseChangeLog:
+  - changeSet:
+      id: 11
+      author: vlasenko
+      changes:
+        - addPrimaryKey:
+            columnNames: users_id, roles_id
+            constraintName: pk_users_roles
+            tableName: users_roles
```
Comment (d1sam) [inline src/main/resources/db/changelog/changes/10-add-primary-key-for-users-roles.yaml:9]:
```suggestion
            tableName: users_roles
            
```

**Review by d1sam — APPROVED** [review]
Well done!

**Pull request merged**